validator/sawtooth_validator/journal/publisher.py

```diff
@@ -32,10 +32,185 @@
+        """
+        txn_hdr = TransactionHeader()
+        txn_hdr.ParseFromString(txn.header)
+        for dep in txn_hdr.dependencies:
+            if dep not in committed_txn_cache:
+                LOGGER.debug("Transaction rejected due " +
+                             "missing dependency, transaction " +
+                             "%s depends on %s", txn.header_signature, dep)
+                return False
+        return True
+
+    def _is_batch_already_committed(self, batch):
+        """ Test if a batch is already committed to the chain or
+        is already in the pending queue.
+        """
+        return (self._block_store.has_batch(batch.header_signature) or
+                batch.header_signature in self._pending_batch_ids)
+
+    def add_batch(self, batch):
+        # first we check if the transaction dependencies are satisfied
+        # The completer should have taken care of making sure all
+        # Batches containing dependent transactions were sent to the
+        # BlockPublisher prior to this Batch. So if there is a missing
+        # dependency this is an error condition and the batch will be
+        # dropped.
+        if self._is_batch_already_committed(batch):
+            # batch is already committed.
+            LOGGER.debug("Dropping previously committed batch: %s",
+                         batch.header_signature)
+            return
+        elif self._check_batch_dependencies(batch, self._committed_txn_cache):
+            self._pending_batches.append(batch)
+            self._pending_batch_ids.add(batch.header_signature)
+            try:
+                self._scheduler.add_batch(batch)
+            except SchedulerError as err:
+                LOGGER.debug("Scheduler error processing batch: %s", err)
+        else:
+            LOGGER.debug("Dropping batch due to missing dependencies: %s",
+                         batch.header_signature)
+
+    def check_publish_block(self):
+        return self._consensus.check_publish_block(
+            self._block_builder.block_header)
+
+    def _sign_block(self, block, identity_signing_key):
+        """ The block should be complete and the final
+        signature from the publishing validator(this validator) needs to
+        be added.
+        """
+        header_bytes = block.block_header.SerializeToString()
+        signature = signing.sign(header_bytes, identity_signing_key)
+        block.set_signature(signature)
+
+    def finalize_block(self, identity_signing_key, pending_batches):
+        self._scheduler.finalize()
+        self._scheduler.complete(block=True)
+
+        # this is a transaction cache to track the transactions committed
+        # up to this batch. Only valid transactions that were processed
+        # by the scheduler are added.
+        committed_txn_cache = TransactionCache(self._block_store)
+        self._committed_txn_cache = TransactionCache(self._block_store)
```

Comment:
I am confused why you create two caches here.  Maybe a comment on why it appears you are basically clearing out the candidate block object's committed transaction cache.